### ISSUE TITLE
concurrency=eventlet leaking?

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#330]
*Originally reported by* **David Szotten (Bitbucket: [davidszotten](https://bitbucket.org/davidszotten), GitHub: [davidszotten](https://github.com/davidszotten))**

----------------------------------------

Been trying to debug a really strange error that appeared when using `coverage --concurrency=eventlet` ("Name or service not known: localhost)

Finally tracked it down to running out of file descriptors, and now that i know what to look for, i can see them build-up (using e.g. `lsof`). I had already noticed memory growing but had chalked that down to book-keeping, but i guess this might be related

are we accidentally preventing gc?

Removing the `--concurrency` flag stops the issue

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/330

**Comment by nedbat** [comment]
*Original comment by* **David Szotten (Bitbucket: [davidszotten](https://bitbucket.org/davidszotten), GitHub: [davidszotten](https://github.com/davidszotten))**

----------------------------------------

think this small example highlights the issue:

```
from weakref import WeakSet
import eventlet


def do():
    pass

gts = WeakSet()
gts.add(eventlet.spawn(do))

print list(gts)
eventlet.sleep(.1)
print list(gts)
```

```
$ coverage run test.py
[<eventlet.greenthread.GreenThread object at 0x1072f3cd0>]
[]


coverage run --concurrency=eventlet test.py
[<eventlet.greenthread.GreenThread object at 0x10e33ee10>]
[<eventlet.greenthread.GreenThread object at 0x10e33ee10>]
```

**Comment by nedbat** [comment]
@davidszotten You are right, the tracer code is holding onto the eventlets too long.  I think I will have to use a WeakRef in the tracing code to be sure that I don't.

**Comment by nedbat** [comment]
Fixed in [2ff50c9162ce (bb)](https://bitbucket.org/ned/coveragepy/commits/2ff50c9162ce)
